File: ios/MassiveRPG/Data/DataSource/UserContextDataSourceImpl.swift
//
//  UserContextDataSourceImpl.swift
//  MassiveRPG-Prod
//
//  Created by Onur Var on 9.04.2024.
//

import Combine
import FirebaseAuth

class UserContextDataSourceImpl: UserContextDataSource, ObservableObject {
    // MARK: Variables

    @Published var user: User?
    var userPublisher: AnyPublisher<User?, Never> { $user.eraseToAnyPublisher() }

    // MARK: Life Cycle

    init() {
        Auth.auth().addStateDidChangeListener { [weak self] _, user in
            guard let user else {
                self?.user = nil
                return
            }
            self?.user = User(
                id: user.uid,
                name: user.displayName,
                email: user.email,
                photoUrl: user.photoURL?.absoluteString
            )
        }

        guard let user = Auth.auth().currentUser else {
            self.user = nil
            return
        }
        self.user = User(
            id: user.uid,
            name: user.displayName,
            email: user.email,
            photoUrl: user.photoURL?.absoluteString
        )
    }
}

File: android/app/src/main/java/com/onurvar/massiverpg/data/datasource/UserContextDataSourceImpl.kt
package com.onurvar.massiverpg.data.datasource

import com.google.firebase.Firebase
import com.google.firebase.auth.auth
import com.onurvar.massiverpg.domain.protocol.datasource.UserContextDataSource

class UserContextDataSourceImpl : UserContextDataSource {
    private val auth = Firebase.auth

    override fun getCurrentUserId(): String? {
        return auth.currentUser?.uid
    }

}